Add support for AWS Redshift clusters and snapshots

herbyg-axial opened the pull request:
This builds off of work by @j-carl in #185 

Changes are as follows:

**redshift** module
  * add the ability to snapshot and restore clusters
  * fix cluster deletion
  * add additional clusters facts, including the cluster SSH public key, node count, VPC ID, tags, etc.
  * add the option to take a final snapshot on cluster deletion
  * add support for current Redshift node types

new module, **redshift_snapshot**
  * list, delete and copy Redshift snapshots
  * ability to list and filter for snapshots by cluster/snapshot name or regular expression, creation time, size, tags and snapshot type
 
@j-carl it looks like you were still making edits to #185, so not sure how best to proceed here.

gregdek commented:
Thanks @herbyg-axial for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

willthames commented on cloud/amazon/redshift.py:
```diff
@@ -0,0 +1,938 @@
+    argument_spec = ec2_argument_spec()
+    argument_spec.update({
+
+        'command':                      dict(choices=['create',
```
Please don't do this. It's horrible to read. Just follow the same formatting as every other module in Ansible

willthames commented on cloud/amazon/redshift_snapshots.py:
```diff
@@ -0,0 +1,860 @@
+    argument_spec = ec2_argument_spec()
+
+    argument_spec.update({
+            'command':                    dict(choices=['list',
```
again, please don't do this.

willthames commented:
Most of my comments on #185 still apply. 

Also, the commits should be squashed to one per author.

:-1: until the module argument_spec gets reformatted at the very least.

herbyg-axial commented:
@willthames I've addressed your comments as well as some other issues.  I will clean up and squash the commits if these latest changes are acceptable.

billwanjohi commented:
I am not a maintainer, but I've been using Ansible and Redshift daily for over two years, and would like to make sure we end up with a quality module.

I think the `command` option here (and all of the control logic that flows from it) misunderstands some core elements of Ansible. Ansible allows us to declare desired state and then ask modules to perform whatever changes (if any!) are needed to achieve that state.

See, for example, the glossary's discussion of [idempotency](http://docs.ansible.com/ansible/glossary.html#idempotency).

See also mentions in the [Developing Modules](http://docs.ansible.com/ansible/developing_modules.html) docs:

> - Do not add list or info state options to an existing module - create a new _facts module.
> - Try and use state present and state absent like other modules

Put simply, options in Ansible modulre should be adjectives, not verbs. Getting more specific:
- `command` option: As the aforementioned documentation states, it would be better to have a `state` option, to conform to the majority of other modules, and to reinforce the declarative, idempotent nature of Ansible.
  - `create` choice: This is not very problematic, but `present` would be better, since after the first run, the module will not in fact be _creating_ anything.
  - `facts`: see aforementioned documentation snippet about avoiding `info` states
  - `delete`: again, not entirely problematic, and can be renamed as `absent` without many implications
  - `modify`: this should be superfluous; if the resource exists, and does not completely match the options specified, it should be modified, if it doesn't, it should be created. This module instead forces the invoker to know the state of all their resources and what ought to be done.
  - `snapshot`: this is totally out of place. While snapshots are actions that are performed on a redshift cluster, they do not actually change the cluster itself. Therefore, in the Ansible context, snapshots should be created, tagged, permissioned, and deleted with a `redshift_snapshot` module, not a `redshift[_cluster]` one. You got most of the way there with redshift_snapshots.py, but the creation should also be taken care of there. I'd also recommend a separate pull request, since each module is plenty complex on its own.
  - `restore`: also superfluous; if a cluster doesn't exist, and a snapshot is specified, one should be created on the basis of that snapshot.

It's also limiting that there is no check mode implemented, and that it uses boto2 (though I understand that the branch predates boto3 general release).

For all of these reasons, I'm working on an alternative module which I hope to submit in the near future, but I thought I'd elaborate on my reasons for those that are still actively developing/reviewing this one.

gregdek commented:
Thanks @herbyg-axial for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@herbyg-axial A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@herbyg-axial Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Looks like this now conflicts with PRs that have been accepted, so closing.
